dashboard/app/views/devise/registrations/new.html.haml

```diff
@@ -94,23 +111,127 @@
     %br/
 
 :javascript
+
+  $(document).ready(function() {
+    // Check for URL having: /users/sign_up?user%5Buser_type%5D=teacher
+    if ("#{resource.teacher?}" == "true") {
+      showTeacher();
+    }
+  });
+
+  function formSuccess(success) {
+    var url;
+    if ("#{return_to_url}" !== "") {
+      url = "#{return_to_url}";
+    } else if (isTeacherSelected()) {
+      url = "#{teacher_dashboard_url}";
+    } else {
+      url = "/";
+    }
+    window.location.href = url;
+  }
+
+  function formError(err) {
+    // Define the fields that can have specific errors attached to them.
+    var fields = ["name", "email", "password", "password_confirmation", "schoolname", "age", "gender"];
+
+    for (var i = 0; i < fields.length; i++) {
+      var field = fields[i];
+      if (err.responseJSON.errors[field]) {
+        var errorField = $("#" + field + "-block .error_in_field");
+        errorField.text(err.responseJSON.errors[field]);
+        errorField.fadeTo("normal", 1);
+      }
+    }
+  }
```

Comment:
I wonder if the UI state here is getting complex enough that it would make sense to user React? 😄